conda/models/prefix_graph.py

```diff
@@ -0,0 +1,387 @@
+                continue
+
+            except StopIteration:
+                return
+
+    @staticmethod
+    def _toposort_pop_key(graph):
+        """
+        Pop an item from the graph that has the fewest parents.
+        In the case of a tie, use the node with the alphabetically-first package name.
+        """
+        node_with_fewest_parents = sorted(
+            (len(parents), node.name, node) for node, parents in iteritems(graph)
+        )[0][2]
+        graph.pop(node_with_fewest_parents)
+
+        for parents in itervalues(graph):
+            parents.discard(node_with_fewest_parents)
+
+        return node_with_fewest_parents
+
+    @staticmethod
+    def _toposort_prepare_graph(graph):
+        # There are currently at least three special cases to be aware of.
+        # 1. The `toposort()` function, called below, contains special case code to remove
+        #    any circular dependency between python and pip.
+        # 2. conda/plan.py has special case code for menuinst
+        #       Always link/unlink menuinst first/last on windows in case a subsequent
+        #       package tries to import it to create/remove a shortcut
+        # 3. On windows, python noarch packages need an implicit dependency on conda added, if
+        #    conda is in the list of packages for the environment.  Python noarch packages
+        #    that have entry points use conda's own conda.exe python entry point binary. If conda
+        #    is going to be updated during an operation, the unlink / link order matters.
+        #    See issue #6057.
+        for node in graph:
+            if node.name == "python":
+                parents = graph[node]
+                for parent in tuple(parents):
+                    if parent.name == 'pip':
+                        parents.remove(parent)
+
+        if on_win:
+            menuinst_node = next((node for node in graph if node.name == 'menuinst'), None)
+            conda_node = next((node for node in graph if node.name == 'conda'), None)
+            python_node = next((node for node in graph if node.name == 'python'), None)
+            if menuinst_node:
+                # add menuinst as a parent if python is a parent and the node
+                # isn't a parent of menuinst
+                assert python_node is not None
+                menuinst_parents = graph[menuinst_node]
+                for node, parents in iteritems(graph):
+                    if python_node in parents and node not in menuinst_parents:
+                        parents.add(menuinst_node)
+
+            if conda_node:
+                # add conda as a parent if python is a parent and node isn't a parent of conda
+                conda_parents = graph[conda_node]
+                for node, parents in iteritems(graph):
+                    if (hasattr(node, 'noarch') and node.noarch == NoarchType.python
+                            and node not in conda_parents):
+                        parents.add(conda_node)
+
+
```

Comment:
This stuff is left over from the old `dag.py` module and should be incorporated later.